### PR TITLE
bugfix: don't initiate ctx directly on the stack

### PR DESCRIPTION
**Pull request by S4mw1s3**
Applied fix as explained here: https://github.com/openssl/openssl/issues/962#issuecomment-208792020
By applying this fix, this module compiles now with OpenSSL 1.1.0c.
Note: I haven't tested if it still compiles with previous versions, but I'm in the assumption it will

### CODE REVIEW

File: src/ngx_http_encrypted_session_cipher.c [hunk 1]
```diff
@@ -26,20 +26,20 @@ ngx_http_encrypted_session_aes_mac_encrypt(ngx_pool_t *pool, ngx_log_t *log,
...
 {
-    EVP_CIPHER_CTX           ctx;
     const EVP_CIPHER        *cipher;
     u_char                  *p, *data;
     int                      ret;
     size_t                   block_size, buf_size, data_size;
     int                      len;
     uint64_t                 expires_time;
     time_t                   now;
+    EVP_CIPHER_CTX          *ctx = EVP_CIPHER_CTX_new();
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: We should avoid using non-constant expressions in C variable declaration statements. This violates the nginx coding disciplines. Also, we need to check if `EVP_CIPHER_CTX_new()` returns `NULL`.

File: src/ngx_http_encrypted_session_cipher.c [hunk 2]
```diff
@@ -142,6 +141,7 @@ ngx_http_encrypted_session_aes_mac_decrypt(ngx_pool_t *pool, ngx_log_t *log,
     const u_char            *digest;
     uint64_t                 expires_time;
     time_t                   now;
+    EVP_CIPHER_CTX          *ctx = EVP_CIPHER_CTX_new();
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: Ditto.

File: src/ngx_http_encrypted_session_cipher.c (hunk 1 above)
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: Also, I wonder if we can avoid allocating a new ctx over and over again on this hot code path by caching it on the global (worker process) level.

**Comment by S4mw1s3** [comment]
I have amended my commit with your requested changes, however caching ctx on the global worker process is something I don't know how to start with. I guess this won't be the same as making ctx a static variable?

File: src/ngx_http_encrypted_session_cipher.c [hunk 3]
```diff
@@ -26,20 +26,25 @@ ngx_http_encrypted_session_aes_mac_encrypt(ngx_pool_t *pool, ngx_log_t *log,
...
     const EVP_CIPHER        *cipher;
     u_char                  *p, *data;
     int                      ret;
     size_t                   block_size, buf_size, data_size;
     int                      len;
     uint64_t                 expires_time;
     time_t                   now;
+    EVP_CIPHER_CTX          *ctx;
+
+    ctx = EVP_CIPHER_CTX_new();
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: This call incurs dynamic memory allocations, which is quite expensive. We should avoid it on this hot code path that is executed in request handlers.

File: src/ngx_http_encrypted_session_cipher.c [hunk 4]
```diff
@@ -142,6 +146,12 @@ ngx_http_encrypted_session_aes_mac_decrypt(ngx_pool_t *pool, ngx_log_t *log,
     const u_char            *digest;
     uint64_t                 expires_time;
     time_t                   now;
+    EVP_CIPHER_CTX          *ctx;
+
+    ctx = EVP_CIPHER_CTX_new();
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: Ditto.

**Comment by agentzh** [comment]
Superseded by #16.

**Pull request closed without merging**